pkg/stats.go

```diff
@@ -8,44 +8,123 @@ import (
+	// including nested types.
+	InputPropertiesMissingDescriptions int
+
+	// TotalOutputProperties is the total number of outputs across all resources, including nested types.
+	TotalOutputProperties int
+
+	// OutputPropertiesMissingDescriptions is the total number of all resource output properties missing descriptions.
+	OutputPropertiesMissingDescriptions int
+}
+
+// FunctionStats contain statistics relating to the functions section of a Pulumi schema.
+type FunctionStats struct {
+	// TotalFunctions is the total number of Pulumi Functions in the schema.
+	TotalFunctions int
+
+	// TotalDescriptionBytes is the sum total of all bytes in the descriptions of the functions themselves,
+	// not including inputs and outputs.
+	TotalDescriptionBytes int
+
+	// TotalInputPropertyDescriptionBytes is the sum total of all bytes in descriptions of function input properties,
+	// not including the input type description.
+	TotalInputPropertyDescriptionBytes int
+
+	// InputPropertiesMissingDescriptions is the total number of all function input properties missing descriptions.
+	InputPropertiesMissingDescriptions int
+
+	// TotalOutputPropertyDescriptionBytes is the sum total of all bytes in description of function output properties,
+	// not include the output type description.
+	TotalOutputPropertyDescriptionBytes int
+
+	// OutputPropertiesMissingDescriptions is the total number of all function output properties missing descriptions.
+	OutputPropertiesMissingDescriptions int
 }
 
 func CountStats(sch schema.PackageSpec) PulumiSchemaStats {
-	stats := PulumiSchemaStats{}
+	stats := PulumiSchemaStats{
+		Resources: ResourceStats{},
+		Functions: FunctionStats{},
+	}
 
 	uniques := codegen.NewStringSet()
 	visitedTypes := codegen.NewStringSet()
 
-	var propCount func(string) (int, int)
-	propCount = func(typeName string) (totalProperties int, propertiesMissingDesc int) {
+	type propCountResult struct {
+		totalInputs        int
+		inputsMissingDesc  int
+		totalOutputs       int
+		outputsMissingDesc int
+	}
+
+	var propCount func(string) propCountResult
+	propCount = func(typeName string) propCountResult {
 		if visitedTypes.Has(typeName) {
-			return 0, 0
+			return propCountResult{}
 		}
 
+		res := propCountResult{}
+
 		visitedTypes.Add(typeName)
 
 		t := sch.Types[typeName]
 
-		totalProperties = len(t.Properties)
-		propertiesMissingDesc = 0
+		res.totalInputs = len(t.Properties)
 
-		for _, p := range t.Properties {
-			if p.Description == "" {
-				propertiesMissingDesc++
+		for _, input := range t.Properties {
+			if input.Description == "" {
+				res.inputsMissingDesc++
 			}
 
-			if p.Ref != "" {
-				tn := strings.TrimPrefix(p.Ref, "#/types/")
-				nestedTotalProps, nestedPropsMissingDesc := propCount(tn)
-				totalProperties += nestedTotalProps
-				propertiesMissingDesc += nestedPropsMissingDesc
+			if input.Ref != "" {
+				tn := strings.TrimPrefix(input.Ref, "#/types/")
+				nestedRes := propCount(tn)
```

Comment:
The lookup into `sch.Types[tn]` on line 88 could fail if the `.Ref` typename is in an external package. That doesn't happen often, but it matters for component providers.

Might want to check here:

```go
if _, has := sch.Types[typeName]; has {				 
  nestedRes := propCount(tn)
  ...
}
```

Reply:
Added a test to ensure this does not panic. @AaronFriel Please resolve the convo if the test correctly addresses this concern, or provide additional feedback if not.